### ISSUE TITLE
Kucoin Too Many Requests problem

### ISSUE THREAD

**Issue by KindSpidey** [ccxt/ccxt#16185]
Hey guys. In my code I ask the info from Kucoin once per 3 secs. It breaks this exchange (while Huobi, Mexc, Bybit, Binance doesnt meet same problems). I got Network Error with
`kucoin GET https://api.kucoin.com/api/v3/market/orderbook/level2?symbol=SDL-USDT 429 Too Many Requests {"code":"429000","msg":"Too Many Requests"}`
Is there any way how to fix it?

**Comment by HuzaifaKhanDeveloper** [comment]
sir can you help me how to run this repo

**Comment by SpenTappy** [comment]
Same issue here. Never had this issue with KuCoin before.

**Comment by KindSpidey** [comment]
@SpenTappy so, you used ccxt and Kucoin before, and nowadays this error starts to happen?

**Comment by carlosmiei** [comment]
Hello @KindSpidey,
Unfortunately kucoin has this chronic problem with the RL, which seems to be random sometimes, can you reproduce the issue deterministically?

**Comment by SpenTappy** [comment]
> @SpenTappy so, you used ccxt and Kucoin before, and nowadays this error starts to happen?

Yes that is correct.  It is not a big problem, we just handle the error.  But I have been noticing it regularly in logs in the last ~10 days .

**Comment by KindSpidey** [comment]
Hi @carlosmiei 
```
while 1:
    print(huobi.fetch_order_book("BTC/USDT")["asks"])
    print(kucoin.fetch_order_book("BTC/USDT")["asks"])
    time.sleep(3)
```
Huobi works fine. Kucoin failed on second operation

**Comment by xmatthias** [comment]
I think the limit posted in this comment still applies (https://github.com/ccxt/ccxt/issues/10273#issuecomment-951078165) 

> orderbooks(lvl2, lvl3), the standard limit is 10req/3s
> their cloudflare limit: 4000/s for a specific endpoint for all users

especially the 2nd one is problematic - as it's nothing you have control over yourself.
we handle this by immediately retrying (only if the error is 429000)  - the 2nd request usually works just fine.

**Comment by KindSpidey** [comment]
@xmatthias that simple? Just to repeat code after getting error?

**Comment by xmatthias** [comment]
in essence - yes.
in practice it'll depend.
So far, i've only seen this error on the ohlcv/klines endpoint. There, we just retry immediately.

If this were to happen on "writing" endpoints (like create order) - then i'd be more concerned, as you can't rely if the order was created or not - but i'm yet to encounter it there - and nobody has reported that so far.

depending on what else you're doing - you might wanna back off slightly (half a second/second) - as you'd be hammering the exchange if this request is the only one you make - potentially leading to different blocks.

That said - ccxt is taking care of this unless rate-limiting is disabled - so it'll depend on your particular usecase to test how it'll actually behave.

**Comment by adamrg73** [comment]
was this fixed in a newer ccxt version, im using an older one because it might break my application to upgrade, but i can run the script on a different server if this was fixed, because im barely making any api requests and I am hitting too many requests all the time.

**Comment by KindSpidey** [comment]
@adamrg73 I believe it cant be solved by ccxt team, it is a Kucoin problem. I talked to developers from Kucoin and they admitted it. Only you can do is retry your command after fail until it's executed

**Comment by KindSpidey** [comment]
> @adamrg73 I believe it cant be solved by ccxt team, it is a Kucoin problem. I talked to developers from Kucoin and they admitted it. Only you can do is retry your command after fail until it's executed

**Comment by kroitor** [comment]
> was this fixed in a newer ccxt version, im using an older one because it might break my application to upgrade, but i can run the script on a different server if this was fixed, because im barely making any api requests and I am hitting too many requests all the time.

As noted above it’s on the exchange side in case of KuCoin - the exchange rate limiter behaves unpredictably at times, so a simple retry strategy in your code might help mitigate the problem.

**Comment by verci-eth** [comment]
guys, I am hitting this issue as well but only after updating ccxt from 2.1.something to the latest one, any take on this?

**Comment by ttodua** [comment]
as 2 years has passed since then, and Kucoin also updated its engine and moved to HTF endpoints, I think this specific issue should be gone as it was.
but let us know if anyone still has any issues with latest CCXT version
